File: series-movies-club-backend/app/services/follow_service.py
"""Business logic for the follower/followee graph."""

from app.extensions import db
from app.models import Follow
from app.utils.error_handlers import APIError


def follow_user(follower, followee):
    if follower.id == followee.id:
        raise APIError("You cannot follow yourself", 400)

    exists = Follow.query.filter_by(
        follower_id=follower.id, followee_id=followee.id
    ).first()
    if exists is not None:
        raise APIError("Already following this user", 409)

    follow = Follow(follower_id=follower.id, followee_id=followee.id)
    db.session.add(follow)
    db.session.commit()
    return follow


def unfollow_user(follower, followee):
    follow = Follow.query.filter_by(
        follower_id=follower.id, followee_id=followee.id
    ).first()
    if follow is None:
        raise APIError("You are not following this user", 404)

    db.session.delete(follow)
    db.session.commit()


def list_followers(user):
    return [f.follower for f in user.followers]


def list_following(user):
    return [f.followee for f in user.following]

File: frontend/src/services/followService.js
import api from './api.js';

export const followUser = async (id) => api.post(`/api/users/${id}/follow`);
export const unfollowUser = async (id) => api.delete(`/api/users/${id}/unfollow`);
export const getFollowers = async (id) => api.get(`/api/users/${id}/followers`);
export const getFollowing = async (id) => api.get(`/api/users/${id}/following`);
